### ISSUE TITLE
Showing server-side error messages in fields

### ISSUE THREAD

**Issue by pardmeister** [icebob/vue-form-generator#236]
Hi

I was wondering how to attach server side validation error messages to input fields. For example, say I have a user model with name surname which I submit to the server via ajax. If the server returns an error for name, ['name': 'Some validation failed', 'surname': 'Some other validation failed'], how do I get this to show under the appropriate fields? I can do this per field by creating a custom validator for each but sometimes need the server to do further validation and let me know which fields are failing with that.

**Comment by lionel-bijaoui** [comment]
Hello,
You should read the documentation about [asynchronous validators](https://icebob.gitbooks.io/vueformgenerator/validation/custom-validators.html#asynchronous-validators).
Good luck

**Comment by pardmeister** [comment]
Hi Lionel

Thanks for the response. I had already implemented an asynchronous validator, and that works well. The way I understand it is that each asynchronous validator is linked to a particular input field. What i was asking is if it was possible to return an error bag from the server for multiple input fields and have vue form validator show those error messages by field name. I am not sure if I am making sense.

I may need to run validation checks for multiple input fields server side, and then hint to the user which fields caused the validation to fail. How would I do that with an asynchronous validator? Is there such an example/use case currently?

**Comment by lionel-bijaoui** [comment]
I'm not sure I'm understanding correctly. I'm going to reopen the issue, in the meantime, try to explain your problem better/differently.
As far as I understand, you want to use the same validator on multiple fields and have each field show the validation message corresponding under each field.
Since, this is how validator works, I don't understand the problem.
What is an "error bag" ? You mean grouping the request into a big one ?

I'm not getting what you want to do here, so please, provide more details, maybe a fiddle if possible.
Thanks.

**Comment by pardmeister** [comment]
Not sure if I can create a fiddle for this use case yet, but let sleep on it. I will try tomorrow morning again. 

In the meantime, I use VFG with Laravel. In the backend, i user laravel's FormRequest to validate the form inputs. In that validation, I may only accept the form if the inputs pass some backend validation which is a combination of the input values and perhaps some database checks. If this validation fails laravel will spew out errors for the offending fields as well as the corresponding error messages. These are in the form of an associative array ['field-name' => 'error-message']. What I wanted to know was if there was a way of attaching the field-name error message to the corresponding field-name in VFG.

I am avoiding doing multiple ajax calls to the server for each of the fields. Let me sleep a bit on this, perhaps I will explain it better tomorrow morning.

**Comment by lionel-bijaoui** [comment]
Ok, I understand better !
It is slipping out of the scope of VFG, but since Laravel is very popular with Vuejs, maybe this need to be explored a bit more.

My opinion is that you should do the multiple ajax call. Trying to avoid that will make everything much more complex. I don't know how Laravel is with API and concurrent call, but it should be fine.

You also got to think about the user experience.
Would you prefer a form that tell you in real time where are your mistakes so you can just correct as you go, or one that tell you when you have finished everything ?

If you still think this is a must have functionality for you app, maybe VFG is not the best choice, or maybe you could do a PR to add this functionality or at least start a solution for this issue ?
If not, don't forget to close the issue if it not relevant anymore.
Good night !

**Comment by pardmeister** [comment]
Thanks, i suppose a good way to learn is to dig into stuff like this. You are right, maybe I should start thinking of a solution for it. It's not a showstopper for my application, I was just wondering if the feature was possible.

**Comment by lionel-bijaoui** [comment]
No problem, it was an interesting idea.
Good luck on your project !

**Comment by vesper8** [comment]
umm.. I think the way Laravel handles it IS the correct way.. while I am not against forms that auto-validate each field in real time by doing an ajax call every time you focus in and out of fields.. that is actually unwanted behaviour for my use case, I would much rather validate the entire form upon the user submitting the form, and have the server validate and return an error bag, as is the case with Laravel's out-of-the-box behaviour

@pardmeister did you figure out a way to do this and make Laravel's default form validation working with VFG by any chance? If so would you mind sharing your solution

**Comment by docmattman** [comment]
I'm also curious to know if this is possible.  Has anyone had any luck validating the the full form server-side in a single request (not each field individually)?

**Comment by zoul0813** [comment]
I think this can probably be achieved currently using a hack, ... let me see if I can articulate my thought process here:

A single custom validator, which just checks an error object that was returned from the server.  On submit, send the form to the server and if it comes back with errors, tell VFG to revalidate the form again, which will trigger the custom validator that reads the server generated errors and returns them.

The custom validator would pass successfully until the forms been validated server side ... this server side validation can be done just after submit, or in intervals as the user completes “groups” of fields, etc.

I can probably whip together a working example of this over the next couple of days, if needed.

However, I think being able to feed VFG an object containing errors would be a good enhancement for v3, as I agree ... validating each field async on blur/change isn’t the best option, especially if your project is using AWS Lambda or API Gateway like mine does ... as I’d be billed for each validation.  I could reduce costs by validating the entire form in one call, and make my app a bit more mobile data friendly as well.

**Comment by zoul0813** [comment]
OK, the example code didn't take very long to do ... [JSFiddle Example](https://jsfiddle.net/ykrtL6sb/6/)

In the example, I just wire up a `setTimeout` that fakes the async ajax call to the server and sets a `customErrorBag` property to the "fake errors from the server".  Then it re-validates the VFG form.

This is a fairly quick/easy hack.  I think a new `setValidationErrors` method can be added to VFG in V3 that would essentially perform a similar task, setting the error array of each component for you which would trigger the display and invalidate the form.  The trick is making sure that your error object needs to mimic your model structure, and the value of each property needs to be an array of strings containing the errors.  You can of course return any value you want from the server, but you'll need to manipulate the server response into the `model.path = ['error1', 'error2']` structure.  Your custom validator can also be written to support whatever your server side error response looks like and return a simple array of error messages as well.

**Comment by jaywilburn** [comment]
i'm trying to do something similar - I have data coming from a service and each element has it's own errorMessage and uses my custom validators. Example: 

`{
        "type": "checkbox",
        "label": "Agreement #5",
        "model": "option5",
        "required": true,
        "errorMessage': "This checkbox is special and must be selected",
        "styleClasses": "col-md-12",
        "validator": "this.checkboxRequired"
      }`
I want to be able to display my own

**Comment by pardmeister** [comment]
@vesper8: Apologies for the late response, I didn't work on it further as it wasn't really a showstopper.

**Comment by dauvray** [comment]
I made this function wich take axios return as argument, hope this can help
```
            handleError(err) {
                this.$refs.child.clearValidationErrors()
                const errors =  err.response.data.errors;
                for ( const error in errors) {
                    this.$refs.child.errors.push({
                        error: errors[error][0],
                        field: _.find(this.schema.fields, {'inputName': error})
                    })
                    let el = document.querySelector(`input[name=${error}]`)
                    el.classList.remove("is-valid")
                    el.classList.add("is-invalid")
                }
            }
```
